src/foam/lib/json/Outputter.java

```diff
@@ -100,56 +108,65 @@ public void outputString(String s) {
     if ( multiLineOutput_ && s.indexOf('\n') >= 0 ) {
       writer_.append("\n");
       writer_.append("\"\"\"");
-      writer_.append(escapeMultiline(s));
+      escape(s, writer_);
       writer_.append("\"\"\"");
     }
     else {
       writer_.append("\"");
-      writer_.append(escape(s));
+      escape(s, writer_);
       writer_.append("\"");
     }
   }
 
-  public String escape(String s) {
-    s = s.replace("\\", "\\\\")
-            .replace("\"", "\\\"")
-            .replace("\t", "\\t")
-            .replace("\r","\\r")
-            .replace("\n","\\n");
-    s = escapeControlCharacters(s);
-    return s;
-  }
-
-  public String escapeMultiline(String s) {
-    s = s.replace("\\", "\\\\");
-    s = escapeControlCharacters(s);
-    return s;
-  }
-
-  public String escapeControlCharacters(String s) {
-    int lastStart = 0;
-    String escapedString = "";
-    char c;
-    for ( int i = 0; i < s.length(); i++ ) {
-      c = s.charAt(i);
-      if ( c >= ' ' ) continue;
-      // Character to hex
-      char right = (char) (c & 0x0F);
-      char left = (char) ((c & 0xF0) >> 4);
-      right += '0';
-      if ( right > '9' ) right += 'A' - '9' - 1;
-      left += '0';
-      if ( left > '9' ) left += 'A' - '9' - 1;
-      char[] escape = new char[] {'\\','u','0','0',left,right};
-      // Add previous string segment
-      escapedString += s.substring(lastStart, i);
-      // Add escape sequence
-      escapedString += new String(escape);
-      lastStart = i + 1;
+  public void escape(String s, Writer writer) {
+    try {
+      char c;
+      int startIndex = 0;
+
+      for ( int i = 0; i < s.length(); i++ ) {
+        c = s.charAt(i);
+
+        String replaceWithCharacter = charactersThatNeedToBeReplaced.get(c);
+        if(replaceWithCharacter != null) {
+          appendSubString(writer, s, startIndex, i);
```

Comment:
Why is this needed? Couldn't you just copy each character as you go rather than remembering ranges and creating substrings?